[숫자야구 미션 다시 풀어보기 리뷰] 리뷰용 PR입니다.

pigpgw opened the pull request:
[숫자야구 미션 다시 풀어보기 리뷰] 리뷰용 PR입니다.

pigpgw committed:
chore: eslint, prettier 설치

pigpgw committed:
docs: 기능 목록 READEME.md 작성

pigpgw committed:
feat: User.js 추가

pigpgw committed:
docs: README.md 고민사항 추가

pigpgw committed:
feat: Random 정답값 생성 기능 추가

pigpgw committed:
feat: add GameMessage.js

pigpgw committed:
add: getUserInput fnc

pigpgw committed:
add: setUserInput fnc

pigpgw committed:
feat: add Validator.js check userInput fnc

pigpgw committed:
fix: validator type check error, game_message import error in validator

pigpgw committed:
feat: add assessStrikeBall fnc

pigpgw committed:
fix: assessUserInput fnc , print

pigpgw committed:
add:replay fnc

pigpgw committed:
chore docs 수정

pigpgw committed:
fix: 비동기 처리 오류 수정 및 validator 사용자 입력 검증 로직 수정

pigpgw committed:
feat: 최종 제출

2yunseong commented on .prettierrc.json:
```diff
@@ -0,0 +1,14 @@
+  "tabWidth": 2,
+  "trailingComma": "es5",
+  "useTabs": false
+}
```
EOL이 있네요

2yunseong commented on package.json:
```diff
@@ -10,7 +10,17 @@
+    "eslint-plugin-prettier": "^5.1.3",
+    "eslint-plugin-react": "^7.33.2",
+    "eslint-plugin-react-hooks": "^4.6.0",
+    "husky": "^9.0.7",
```
오.. husky까지

2yunseong commented on src/Validator.js:
```diff
@@ -0,0 +1,28 @@
+    isDuplicate(input){
+        const set = new Set(input)
+        const newArr = [...set]
+        if (newArr.length !== 3) throw new Error(GameMessage.INVALID_INPUT_IS_DUPLICATED);
```
3이라는 숫자는 어떤 의미를 가지고 있는 숫자 같은데요,
이러한 숫자가 숫자 그대로 존재하는 것을 흔히 매직넘버라고 부릅니다.
따라서 이러한 숫자는 다음의 이유로 상수화 해주는 것이 좋습니다.

1. 예시로 숫자야구 게임의 숫자갯수가 4로 늘어났다. 그렇다면 모든 파일을 뒤져서 3을 4로 고쳐야한다. (변경에 용이하지 않다.)
2. 처음 온보딩 하는 사람은 이 숫자의 의미를 파악하기 힘들다. 코드 가독성이 줄어든다.

2yunseong commented on src/Validator.js:
```diff
@@ -0,0 +1,28 @@
+
+    isDuplicate(input){
+        const set = new Set(input)
+        const newArr = [...set]
```
1. 코드를 다음과 같이 고치면 더 줄어듭니다 ㅎㅎ
```suggestion
        const newArr = [...new Set(input)]
```
2. newArr이라는 상수명이 어떤 의미를 담고 있는 이름인가요? 의미가 없고, 단발성으로 사용되는 측면에서
이렇게 작성해도 될 것 같아요.
```js
isDuplicate(input) {
  if ([...new Set(input)].length !== 3) throw new Error(GameMessage.INVALID_INPUT_IS_DUPLICATED)
  return true;
}
```

2yunseong commented on src/Validator.js:
```diff
@@ -0,0 +1,28 @@
+    isNumber(input){
+        if (isNaN(input)) throw new Error(GameMessage.INVALID_INPUT_NOT_A_NUMBER)
+        return true
+    }
```
1. 개인적인 견해로는 검증함수에서는 해당 input이 유효한지/안한지(boolean) 평가한 다음,
이 결과에 따라 에러를 던져주는 패턴을 선호합니다.

다시 말하면 에러를 던져주는 부분을 이 함수(is..)를 호출하는 부분으로 넘기는 방식으로 하곤합니다

2. 이게 static method가 아닌 method일 이유가 있나요?

2yunseong commented on src/Validator.js:
```diff
@@ -0,0 +1,28 @@
+        if (isNaN(input)) throw new Error(GameMessage.INVALID_INPUT_NOT_A_NUMBER)
+        return true
+    }
+    // 먼저 길이 체크 후 중복 검사
```
필요없는 주석은 지우면 좋겠습니다 !

2yunseong commented on src/Validator.js:
```diff
@@ -0,0 +1,28 @@
+import GameMessage  from "./GameMessage.js";
+class Validator {
+    validInput(input){
+        this.isNumber(input)
+        this.isValidLength(input)
+        this.isDuplicate(input)
+        return true
+    }
```
boolean을 리턴할 때는 메서드의 prefix에 is, has를 붙여주는 것이 관례입니다ㅎ_ㅎ

2yunseong commented on src/Computer.js:
```diff
@@ -0,0 +1,35 @@
+    }
+
+    assessUserInput(input) {
+        let strike = 0;
+        let ball = 0;
```
어떤 구문에는 세미콜론이 있고, 어떤 구문에는 세미콜론이 있지 않네요.

둘 중 하나만 하시는 걸 추천드리고, 강제하고 싶다면 prettier등을 이용해보세요.

2yunseong commented on src/Computer.js:
```diff
@@ -0,0 +1,35 @@
+    assessUserInput(input) {
+        let strike = 0;
+        let ball = 0;
+        const userInput = input.split("").map(item => Number(item));
```
1. 이렇게도 됩니다.
```suggestion
        const userInput = input.split("").map(Number);
```
2. 배열이면.. s가 붙으면 좋지 않을까요?

2yunseong commented on src/Computer.js:
```diff
@@ -0,0 +1,35 @@
+        let strike = 0;
+        let ball = 0;
+        const userInput = input.split("").map(item => Number(item));
+        userInput.forEach((item, index) => {
+            if ((this.#solution.indexOf(item) === index) && this.#solution.includes(item)) strike += 1;
+            else if ((this.#solution.indexOf(item) !== index) && this.#solution.includes(item)) ball += 1;
+        });
```
reduce를 써서도 구현할 수 있겠네요
```js
const strikeAndBall = userInput.reduce((acc, cur) => {
  if ((this.#solution.indexOf(item) === index) && this.#solution.includes(item)) return { ...acc, strike: acc.strike + 1 }
if ((this.#solution.indexOf(item) !== index) && this.#solution.includes(item)) return { ...acc, ball:acc.ball + 1}
return acc;
}, { strike: 0, ball: 0})
```

2yunseong commented on .prettierrc.json:
코드를 보면 prettier가 적용되어있지 않은데, 확인한번 해보세용

2yunseong commented on src/Computer.js:
```diff
@@ -0,0 +1,35 @@
+                computer.push(number);
+            }
+        }
+        return this.#solution = computer
+    }
```
이걸 리턴하는 이유가 뭔가요 ?

2yunseong commented on src/App.js:
```diff
@@ -1,5 +1,52 @@
+    }
+  }
+
+async setUserInput(){
+    const input = await Console.readLineAsync("숫자를 입력해주세요 : ");
+  if (this.validator.validInput(input)) {
+    this.user.setNumber(input);
+    }
+}
```
indent가..

2yunseong reviewed (changes requested):
아직까지는 App이 가지고 있는 역할과 책임이 커보여요
조금 분산해보시길...

고생하셨습니다 ㅎ_ㅎ

pigpgw committed:
피드백 1차 반영

pigpgw committed:
fix: 피드백 1차 반영

pigpgw committed:
fix: 피드백 전체 반영

pigpgw committed:
최종 제출

pigpgw committed:
Refactor Validator class in Validator.js

- Fixed a logical error in the isValidInput method where isValidLength was called twice.
- Updated isValidLength to return a boolean instead of throwing an error when the length check fails.
- Updated isNumber method to return a boolean directly.
- Updated isDuplicate method to return a boolean directly.

pigpgw committed:
Apply 'static' keyword to methods in Validator class

- Applied the 'static' keyword to the methods in the Validator class to denote that they are utility methods and do not require an instance of the class to be invoked.
- By making these methods static, they can be called directly on the class itself, improving code organization and making it clearer that they are standalone utility functions related to validation.
- This change enhances the flexibility and usability of the Validator class, allowing these methods to be conveniently accessed without the need to instantiate the class.

pigpgw committed:
Fix ESLint issues

- Updated codebase to adhere to ESLint rules and standards.
- Addressed various linting errors and warODnings, including:
  - Corrected indentation inconsistencies.

pigpgw committed:
add endOfLine: auto to Prettier configuration

2yunseong commented on src/App.js:
```diff
@@ -1,5 +1,64 @@
+  async mainLogic() {
+    await this.setUserInput();
+    this.printStrikeBall();
+    if (this.computer.assessUserInput(this.user.getNumber()).strike !== 3)
```
매직넘버가 살아이있네용

pigpgw committed:
fix: Change FULL_STRIKE magic number to 3

pigpgw committed:
refactor: MIN_RANDOM_NUMBER MAX_RANDOM_NUMBER masic number to 1 , 9

2yunseong commented on src/MasicNumber.js:
정녕 MagicNumber라는 이름이 최선인가요?

2yunseong commented on src/App.js:
```diff
@@ -1,5 +1,67 @@
+// eslint-disable-next-line import/no-unresolved
+import { Console } from "@woowacourse/mission-utils";
+// eslint-disable-next-line import/no-unresolved
+import User from "./User.js";
+// eslint-disable-next-line import/no-unresolved
+import Computer from "./Computer.js";
+// eslint-disable-next-line import/no-unresolved
+import Validator from "./Validator.js";
+// eslint-disable-next-line import/no-unresolved
+import GameMessage from "./GameMessage.js";
+// eslint-disable-next-line import/no-unresolved
+import { RETRY_RESPONSE, FULL_STRIKE } from "./MasicNumber.js";
```
이렇게 eslint 규칙을 무시할거면 eslint를 왜 사용하시나요?

pigpgw committed:
fix: print strike ball message error and apply code review
